### ISSUE TITLE
Simplify CMake

### ISSUE THREAD

**Issue by Psirus** [nutofem/nuto#98]
Having recently touched almost all CMake files, I've noticed that there are still a lot of unnecessary complications in there. Maybe we can simplify our build system a bit.

# External libraries and static linking

## BUILD_SHARED_LIBS

Does anyone need NuTo as a static lib? For that, you would need to have all the external libs as static libs as well. The size of your executables becomes huge. I don't think this is necessary anymore.

## ENABLE_STATIC_EXTERNALS

This is connected to the previous point: We get our libraries primarily from the Debian repos, compiling them yourself just isn't fun. Also, I think we're at a point where introducing a dependency that is not in the Debian repos or a single header we can put in `external/`, this will meet a lot of resistance. I'd say this should just go as well.

## NUTO_EXTERNAL_LIBRARIES_ROOT

Again, same difference. If you don't expect to find your libs in unusual places, you don't need this variable. Let's get rid of it.

# Preprocessor definitions

**Don't confuse with CMake options of the same name.**

Preprocessor definitions, aka.
~~~cpp
void MyFunction()
{
#ifdef ENABLE_OPTIMIZE
    doThis();
#else
    doThat();
#endif
}
~~~
are very bad for readability and reasoning about the code. They should be kept to a minimum. `ENABLE_VISUALIZE` will probably stay for a while, but `ENABLE_MECHANICS`, `ENABLE_OPTIMIZE` should go. What about `HAVE_MUMPS`, `HAVE_PARDISO`, `HAVE_ARPACK`?

---

- [x] Upgrade to CMake 3.5
- [x] Use Boost::filesystem syntax
- [x] ENABLE_STATIC_EXTERNALS
- [x] NUTO_EXTERNAL_LIBRARIES_ROOT
- [x] ENABLE_MECHANICS
- [x] ENABLE_OPTIMIZE
- [x] ENABLE_VISUALIZE
- [x] ENABLE_METAMODEL
- [x] ENABLE_GEOMETRYCONCRETE
- [x] ENABLE_ARPACK
    - [ ] HAVE_ARPACK?
- [x] `ENABLE_TESTS` + `make test`
- [x] `ENABLE_COVERAGE`
- [x] `ENABLE_EXAMPLES` + `make examples`
- [x] `ENABLE_BENCHMARKS` + `make benchmarks`
- [x] Remove FindOPENBLAS
- [x] Remove FIND_STATIC_LIBRARY
- [x] Move dependency finding to seperate file
- [x] Update "our" FindXYZ to IMPORTED targets syntax as well
- [x] remove Boost::system, not used anymore (I think)
- [x] remove `deactivate_module`
- [x] some refactoring
- [ ]  get rid of "raw" `include_directories` (see comments below)
- [ ] `ENABLE_MPI`?
- [ ] `ENABLE_OpenMP`?
- [ ] optional solvers -> don't force one solver, just make sure that one is found
- [x] `ENABLE_CUSTOM_EXAMPLES`?
- [ ] better handle benchmarks, see also #42 
- [ ] remove linking examples to boost unit test, see also #75 
- [ ] dependency integrationtests <-> python
- [x] QTCreator specific code needed?
- [ ] Imported targets for Eigen, MKL, Pardiso

**Comment by TTitscher** [comment]
/agreed.

MUMPS and ARPACK are included in the debian repo. IMO we can assume that they are available and remove this flag. Pardiso requires registration and an external lib, so we should keep `HAVE_PARDISO`.

As for the internal `ENABLE_...` flags. They do not depend on any special outside libs. So there is no need to disable e.g. visualize. If you do not need it, turn it off at runtime.

**Comment by Psirus** [comment]
Wrt. to the `ENABLE_ARPACK`, `ENABLE_MUMPS` and `ENABLE_PARDISO`: I think they should be "optional", i.e. always look for them, if found use them, if not found warn that they're missing and deactivate the respective parts.

**Comment by Psirus** [comment]
Starting from CMake 3.0, there is this concept of imported targets. You may remember
`target_link_libraries(MyExecutable NuTo::Mechanics)`. Here, `NuTo::Mechanics` is such an imported target. The same could be done for other targets. Sadly, FindBoost exports such targets only from CMake 3.5 onwards. Using this would have a couple of advantages:
- the syntax is nicer: `Boost::filesystem` instead of `${Boost_FILESYSTEM_LIBRARY}`
- "config-time-checking" rather than at link time: `Boost:filesysetm` will get you a CMake error, `${Boost_FILESYSETM_LIBRARY}` will get you a linker error
- you don't need to specify the include directories separately any more; if a target links to Boost, it also gets its includes

In Debian stable, CMake is 3.7, in oldstable 3.0 (backports 3.6), and Travis is 3.2. Do we want this functionality, and then set the minimum version to 3.5 and update CMake on Travis?

Edit: [This pull requests](https://github.com/travis-ci/travis-cookbooks/pull/891) adds CMake 3.9 to Travis. If this gets merged, we don't have to manually install CMake.

**Comment by TTitscher** [comment]
Since cmake>3.5 is in both debian-stable and ubuntu16.04LTS, I vote for an upgrade to cmake3.5. Then, it would be nice, if we manage to get version 3.5 on travis, instead of 3.9. To detect, if someone accidentally uses features >3.5...

**Comment by TTitscher** [comment]
In our current CMakeLists, we have a lot of manual install scripts. Mainly, because travis only supports ubuntu trusty (14.04) out of the box and we decided to rely on the next LTS version ubuntu 16.04. Examples are _install_cmake.sh_, _install_numpy.sh_, _install_gmsh.sh_. Then, we have some special cases for old version, e.g. SWIG2.0 that does not support enums. 

This all seems kind of hacky and for a proper cleanup of our CMake system, it might be advantageous to approach this docker thing (with ubuntu 16.04) once again (#64).

**Comment by Psirus** [comment]
Can I get an opinion on some of the `ENABLE_XYZ` things? Imho these should stay:
- `ENABLE_BENCHMARK`
- `ENABLE_TESTS`
- `ENABLE_COVERAGE`
- `ENABLE_EXAMPLES`
- `ENABLE_CUSTOM_EXAMPLES`
- `ENABLE_PYTHON`

As @TTitscher has said, our own modules should just be `ON` all the time, I don't see the harm. That would remove:
- `ENABLE_GEOMETRYCONCRETE`
- `ENABLE_METAMODEL`
- `ENABLE_OPTIMIZE`
- `ENABLE_VISUALIZE`

Then there are `ENABLE_OPENMP` and `ENABLE_MPI`; don't know with those. Just leave it?

The last group are external libraries, where I would prefer "look for them, and if not found, disable that functionality". But I'm not sure how everyone is using these w.r.t. "choose at config time, choose at compile time, choose at runtime" etc. Maybe you can comment on how you would like to handle these cases, especially those of you who are using them: (btw: do these still work? can we test that?)
- `ENABLE_MKL` 
- `ENABLE_PARDISO`

**Comment by TTitscher** [comment]
My suggestion to aggressively reduce options:
- enable our code by default. No more ENABLE_OPTIMIZE...
- control our targets via make targets. `make test`, `make benchmark`, `make unit`. No need to have ENABLE_BENCHMARK. (Maybe I am missing something...)
- could we even have a `make python_wrapper` target? No ENABLE_PYTHON
- require 'repository dependencies', like we do with boost, eigen, ann, ... No more ENABLE_MUMPS.
- enable OpenMP by default, use it with NumCores=1 if you want a serial run
- replace ENABLE_COVERAGE with the compiler flag `--coverage` in the corresponding travis run.
- in the long run: Use NuTo as a lib, which makes ENABLE_CUSTOM_EXAMPLES obsolete. This maybe even reduce external deps. If you want to use the PardisoSolver, you have to link the lib to your custom mainfile.

For now, we could end up with those options.
~~~cmake
option(ENABLE_MKL "enables support for mkl blas, lapack, solvers" FALSE)
option(ENABLE_PARDISO "enables support for PARDISO solver" FALSE)
option(ENABLE_MPI "enables message passing interface" FALSE)
option(ENABLE_CUSTOM_EXAMPLES
    "enable private examples in separate directory (applications/custom)" FALSE) # removed if LibNuT
option(ENABLE_PYTHON "create python wrapper for nuto" TRUE) # maybe via `make python_wrapper` target
~~~

__Edit__. What about MPI? (Totally no expert here...) Can we use this option by default? If we don't run our examples with `mpirun`, we have a normal run, right?

**Comment by Psirus** [comment]
I like the idea of using the build targets for tests etc. With Python I'm afraid this won't be possible. If you want to not build a target during an unqualified `make`, you will have to set `EXCLUDE_FROM_ALL` (we'll have to set this for the tests and benchmarks). However, CMake says:

> Installing a target with EXCLUDE_FROM_ALL set to true has undefined behavior.

which I understand to mean we can't install Python and `EXCLUDE_FROM_ALL` at the same time, which makes sense. So `ENABLE_PYTHON` stays. Unless we want to make it mandatory for everyone. (Or get rid of it completely).

I don't think we should "require" things that aren't required (haha). Boost and Eigen are required imho, there is no way to build NuTo without them. Same probably goes for ANN. But MUMPS is not really required, you should be able to choose your own solver. Same with Arpack, if you don't want to solve eigenvalue problems, don't install it. This might incur a small overhead in our CMake code, but if you ever want to install it on a system that is not Debian/Ubuntu, or that you don't have root access to, this becomes much easier. That doesn't mean that we have to keep `ENABLE_MUMPS`. Ideally, I would just check for MUMPS, PARDISO, PETSc, whatever, and only error if no solver is found.

I agree on the custom examples. Not sure yet with OpenMP and MPI.

**Comment by joergfunger** [comment]
I agree with the idea from @TTitscher . OpenMP can always be turned on. Regarding MPI, we should talk to @phuschke , since he is currently the only one using it. I'd like to postpone the decision on this specific flag. Regarding Pardiso, I would opt for combining MKL and Pardiso into a single flag and only use the intel MKL Pardiso solver.

**Comment by Psirus** [comment]
TL;DR: **Doing convoluted shit in CMake leads to more convoluted shit.**

"Modern CMake" relies on the notion of targets and their properties. For example, you could have an imported `Mumps::Mumps` target. When you want to build an executable that links to it, you can just say

    add_executable(MyApp someSource.cpp moreSource.cpp moreSauce.cpp)
    target_link_libraries(MyApp Mumps::Mumps)

If you had used the old `${MUMPS_LIBRARIES}` syntax, you would now also set the include directories if you need one of the Mumps headers, `target_include_directories(MyApp ${MUMPS_INCLUDE_DIR})`. With the new targets, this is not necessary anymore. CMake knows that the headers are necessary, and includes them when you say `target_link_libraries(MyApp Mumps::Mumps)`. So far so good.

Now, another feature of CMake that we are using are OBJECT libraries. This creates individual targets for object files. What are we using these for? Normally, like in the example above, you just specifiy the source files directly. However when you have targets in different directories, e.g.

    # in src/math
    add_library(Math someMath.cpp src/base/someBase.cpp)

    # in src/mechanics
    add_library(Mechanics someMechanics.cpp src/base/someBase.cpp)

and they are using the same source file (here `src/base/someBase.cpp`), this file would be compiled twice, because the flags you use for `Math` and the flags you use for `Mechanics` could be different.

Often when making a change, you want to know if it still works (run your unit test) before changing all the places where the code is used. Therefore, for building our unit tests we wanted to avoid linking to the NuTo libraries, because that would mean you cannot run them without getting the whole library to build. But at the same time, building every source file twice, once for the lib and once for the unit test, would incur a compile-time penality we were not willing to pay. Insert the OBJECT libraries.

Basically, all source files in NuTo are compiled to the OBJECT libraries before being linked to the respective targets.

    # in NutoMacros
    add_library(math.Legendre OBJECT math/Legendre.cpp)
    # in src/math
    add_library(math $<TARGET_OBJECTS:math.Legendre>)
    # in test/math
    add_executable(LegendreTest $<TARGET_OBJECTS:math.Legendre>)

Again, so far so good. But there is a problem with the combination of the two.

If I know say `target_link_libraries(math Mumps::Mumps)`, the include directories of the `math.Legendre` target will not include the Mumps include directories, and if it needs them, it obviously will not compile. 

So what to do? I see a couple of options:

a) Don't properly use the new features and just keep piling on the include directories for all the targets as we have been doing. Meh.
b) Now that `make unit` and `make` don't overlap anymore, is it acceptable to build things twice?
c) Expand the functionality of our CMake macros. There are a chain of macros that gets you from the list of sources to the finished NuTo library, to be found in `scripts/NuToMacros.cmake` (specifically `create_nuto_module`, `source_to_objects`, `create_object_lib`). I'm sure these could be augmented to pass the include directories through to the OBJECT libraries. But that would have to be written and maintained as well.

Your thoughts?

**Comment by joergfunger** [comment]
I would opt for option b). The workflow is usually separated in that you would either require to compile unit tests(in the development) or the lib (when trying to set up your examples). As a consequence, the overhead in terms of compile time is not that much of an issue. Option c) is not my preferred option, since I would prefer to have an interface that is easy to understand - in particular when looking at the core group of NuTo developers that is constantly changing (students joining the group and others leaving).

**Comment by Psirus** [comment]
Assuming that #116 is accepted, many of the things discussed here will be implemented. As a to-do list for the next person wanting to tackle this, here are the things I didn't get to:
- [ ] Remove all raw `include_directories()`, related to the comment above.
- [ ] How to handle the benchmarks? Should there be a `benchmarks` target? Should they be part of the test? Related: #42 
- [ ] As mentioned multiple times: examples should not link to Boost unit test. Solve #75.
- [ ] The integration tests require the Python wrappers, but the `integrationtests` target does not yet depend on them. Fix needed.
- [x] There are a couple of lines in the CMakeLists.txt: "This makes qtcreator show header-only classes". Imho that should not be there. If it is genuinely useful to QT Creator users, fine. But maybe it isn't needed anymore; can someone check with a current QT Creator version? Maybe there is another way to tell QT Creator about the headers?
- [ ] Imported targets for
    - [x] Eigen (btw, I think `FindEigen` should return the include path including the `eigen3` prefix, so as to make the `#include <Eigen/Core>` equivalent to the Eigen documentation)
    - [ ] Pardiso, MKL, Lapack; for this it would be really helpful to have testing, maybe on Jenkins?
- [ ] What to do with `ENABLE_OPENMP` and `ENABLE_MPI`.

For now, I'm contend with what is done and will leave it at that.

**Comment by vhirtham** [comment]
Can we add a build target that simply builds all the enabled stuff without the need to call make *whatever*? The current implementation conflicts with my IDE in several ways and it is quiet a pain to circumvent the issues.

**Comment by Psirus** [comment]
You can add a build target that depends on all the other targets and therefore builds everything, maybe call it `everything` :) But you would still need to call "make *whatever*", namely `make everything`.

**Comment by vhirtham** [comment]
Maybe I just name it `all_targets` ? Specifying a single target is not the problem at all.

**Comment by Psirus** [comment]
I think that it could cause confusion between the `all` target, which does something different. But in the end, not really important.

**Comment by joergfunger** [comment]
What exactly is the problem with your build environment?

**Comment by vhirtham** [comment]
Problem is that you should enjoy your vacations instead of being here :p 

Serious answer:

If I build the project in my IDE in order to start debugging a testfile, the test is not build when the target is `all`. So I switch to `test` or  `unit` and try it again. 
Let's say  there is an error in my file, which was the reason to start the debugger in the first place. The build finishes without errors but immediately afterwards the tests are started as if I typed `make test` or `make unit`. My test fails during the test run and causes my IDE to think the build has failed. As a result, the debugging session won't start. 

Maybe there is an option in the IDE to suppress the tests, but I have not found it yet. I can select a single target too, but it would be a total pita to change that every time.

**Comment by Psirus** [comment]
So your problem is not actually the lack of a target that builds everything, but that the tests are also run, not only build. In 5c5eafa1 I've deactivated that for the integrationtests and examples, so this should no longer happen for `make integrationtests`. Doing the same for the unit tests would be straightforward.

**Comment by vhirtham** [comment]
Didn't know that this was actually an discuss-able option. However I found another option in QT Creator. There is a flag `make <current executable>`. So I do this after `make all` and the initial problem should be gone. Sorry for bothering you.

**Comment by TTitscher** [comment]
Glad that you found this option. And generally: Everything is discuss-able!
